### PR TITLE
Data sets

### LINKED ISSUE #2: Data loading
Simple approach: store dataset in numpy arrays in memory.

- [x] Mnist
- [x] Fashion-Mnist

Three classes:
- `AbstractDataset()`
- `Mnist()`
- `FashionMnist()`

Required arguments:
- `batch_size`

Optional arguments:
- `shuffle`
- `repeat`

### PR DESCRIPTION
**Pull request by Jopyth**
Fixes #2.

### CODE REVIEW

File: length/data_sets/mnist_like.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            labels = label_array[self.batch_size * i:self.batch_size * (i + 1)]
+            yield Batch(data, labels)
+
+    def prepare(self):
+        self.download_files()
+        self.load_files_into_memory()
+        self.check_sanity()
+
+    def download_files(self):
+        if not os.path.exists(self.path):
```
Comment (Bartzi) [inline length/data_sets/mnist_like.py]: you could also write `os.makedirs(self.path, exist_ok=True)`

File: length/data_sets/mnist_like.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            os.makedirs(self.path)
+
+        urls = [self.data_url + f for f in self.basenames]
+        for url, basename in zip(urls, self.basenames):
+            output_file_name = os.path.join(self.path, basename)
+            if not os.path.isfile(output_file_name):
+                print("downloading file %s..." % basename)
+                urlretrieve(url, output_file_name)
```
Comment (Bartzi) [inline length/data_sets/mnist_like.py:70]: a progress bar would be cool^^ (just kidding)

File: length/data_sets/mnist_like.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    def load_files_into_memory(self):
+        for file, target in self.basenames.items():
```
Comment (Bartzi) [inline length/data_sets/mnist_like.py]: I'd suggest a different name than `file` because `file` shadows the built in `file` function. How about `file_name`?

File: length/data_sets/mnist_like.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            file_path = os.path.join(self.path, file)
+            with gzip.open(file_path) as handle:
+                magic, = struct.unpack(">I", handle.read(4))
```
Comment (Bartzi) [inline length/data_sets/mnist_like.py]: a link to the dataset specification would be cool here :smile:

File: length/data_sets/mnist_like.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                binary_magic = format(magic, '08X')
+                assert binary_magic[:4] == "0000"
+                assert binary_magic[4:6] == "08"
+                dimensions = int(binary_magic[6:], 16)
+                shape = struct.unpack(">" + "I" * dimensions, handle.read(4 * dimensions))
+                data = np.fromstring(handle.read(), dtype=np.int8)
+                if dimensions > self.dimensions:
+                    shape = shape[0:1] + (int(np.prod(shape[1:])),)
+                if dimensions < self.dimensions:
+                    shape = shape[0] + [1] + [shape[1:]]
```
Comment (Bartzi) [inline length/data_sets/mnist_like.py]: this would fail, right? it should be `shape[0:1] + (1,) + shape[1:]`

File: train.py [hunk 6]
```diff
@@ -31,6 +31,9 @@ def main(args):
 
             loss.backward(optimizer)
 
+            if iteration % 50 == 49:
```
Comment (Bartzi) [inline train.py]: I suggest setting this to `iteration % 50 == 0`^^

**Review by Bartzi — CHANGES_REQUESTED** [review]
please have a look at my comments ;)

**Review by Bartzi — APPROVED** [review]
:sun_with_face:
